### PR TITLE
fix: unable to reference multiple structs that share a dotwise base name

### PR DESCRIPTION
**Pull request by iliapolo**
As mentioned in this [comment](https://github.com/cdk8s-team/cdk8s-cli/pull/54#issuecomment-899598963), the underlying problem was that when types are created from references, the type name used to only be the last element of a possibly dotwise namespace.

This reduced references like `io.k8s.v1beta.Foo` and `io.k8s.v1.Foo` to resolve to the same type name (i.e `Foo`), causing only the first of which to be emitted and referenced to. 

This PR applies a normalization logic to those references. By default, dotwise namespaces are normalized by pascal casing each element and joining them together, just as we already do for property types:

https://github.com/cdklabs/json2jsii/blob/0ddf8e1448d891c5f8d9173592372f79c6698651/src/type-generator.ts#L484-L486

This will result in interfaces named `IoK8Sv1Beta1Foo` - which might be a little too verbose. To that end, we also provide a way to pass a custom ref normalizer.

Fixes https://github.com/cdk8s-team/cdk8s-cli/issues/45

### CODE REVIEW

File: src/type-generator.ts [hunk 1]
```diff
@@ -496,8 +504,7 @@ export class TypeGenerator {
       return { type: 'any', toJson: x => x };
     }
 
-    const comps = def.$ref.substring(prefix.length).split('.');
-    const typeName = TypeGenerator.normalizeTypeName(comps[comps.length - 1]);
```
Comment (iliapolo) [inline src/type-generator.ts:500]: This was the offending code. Instead, we consider the entire namespace and normalize it.

File: src/type-generator.ts [hunk 2]
```diff
@@ -33,6 +34,11 @@ export interface TypeGeneratorOptions {
    * @default true
    */
   readonly toJson?: boolean;
+
+  /**
+   * Normalization function to apply to references which are used to create
+   */
+  readonly normalizeRef?: (s: string) => string;
```
Comment (iliapolo) [inline src/type-generator.ts:43]: Not in love with this. But I thought we would need this in `cdk8s` so we don't end up with horrible looking interface names.

Comment (eladb) [reply src/type-generator.ts:43]: I think I would call this `renderTypeName(def: string) => string`

File: src/type-generator.ts [hunk 3]
```diff
@@ -546,6 +555,11 @@ export class TypeGenerator {
 
     return false;
   }
+
+}
+
+function normalize(s: string, pre: (s: string) => string) {
```
Comment (eladb) [inline src/type-generator.ts:561]:
```suggestion
function normalize(s: string, pre: (s: string) => string = DEFAULT) {
```

File: src/type-generator.ts [hunk 4]
```diff
@@ -546,6 +555,11 @@ export class TypeGenerator {
...
+  return TypeGenerator.normalizeTypeName(pre(s));
```
Comment (eladb) [inline src/type-generator.ts:562]: Consider just calling `pre` outside of this function

File: src/type-generator.ts (hunk 2 above)
Comment (iliapolo) [reply src/type-generator.ts]: Changed

File: src/type-generator.ts (hunk 3 above)
Comment (iliapolo) [reply src/type-generator.ts]: Got rid of this function altogether

**Review by eladb — APPROVED** [review]

**Pull request merged**